google/cloud/spanner/internal/compiler_info.cc

```diff
@@ -77,6 +77,29 @@ std::string CompilerVersion() {
   return "Unknown";
 }
 
+std::string CompilerFeatures() {
+#if GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
+  return "ex";
+#else
+  return "noex";
+#endif  // GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
+}
+
+std::string LanguageVersion() {
+  switch (__cplusplus) {
+    case 199711L:
+      return "98";
```

Comment:
Your changes to this function to remove the function static std::string are better, thanks. The std::string was originally needed when a previous version of this code concatenated strings like "ex-" with the 4-digit year. But as that functionality was split up, the function wasn't simplified the rest of the way. So, thanks for this cleanup.

FWIW, the previous code was not actually leaking anything. You've probably heard about C++'s static initialization-order fiasco? That's not really a problem with function-local statics. A related but harder-to-control problem is the lesser known static destruction fiasco. That is, static data is destroyed in some order on program shut down, but in an environment with (potentially) multiple threads, it's *very* difficult to ensure that no threads access the static data *after* it has been destroyed.

A common and easy fix to this problem is to never destroy static objects. This is accomplished by heap-allocating the object, holding a function-static *pointer* to the object, and never deleting it. This is not a leak because there will always be a valid pointer pointing to the heap-allocated object. Leak detectors will not (or should not, else I think it's their bug) report this as a leak.

Reply:
Thanks for the info and search terms for me to research later. The more I learn about C++ the better.